[MLIR][Bufferization] BufferResultsToOutParams: Add an option to eliminate AllocOp and avoid Copy

Menooker opened the pull request:
Add an option hoist-static-allocs to remove the unnecessary memref.alloc and memref.copy after this pass, when the memref in ReturnOp is allocated by memref.alloc and is statically shaped. Instead, it replaces the uses of the allocated memref with the memref in the out argument.
By default, BufferResultsToOutParams will result in a memcpy operation to copy the originally returned memref to the output argument memref. This is inefficient when the source of memcpy (the returned memref in the original ReturnOp) is from a local AllocOp. The pass can use the output argument memref to replace the locally allocated memref for better performance.hoist-static-allocs avoids dynamic allocation and memory movement.
This option will be critical for performance-sensivtive applications, which require BufferResultsToOutParams pass for a caller-owned output buffer calling convension.

wyzero commented on mlir/lib/Dialect/Bufferization/Transforms/BufferResultsToOutParams.cpp:
```diff
@@ -118,10 +119,14 @@ static LogicalResult updateReturnOps(func::FuncOp func,
-              memCpyFn(builder, op.getLoc(), std::get<0>(t), std::get<1>(t))))
-        return WalkResult::interrupt();
+    for (auto [orig, arg] : llvm::zip(copyIntoOutParams, appendedEntryArgs)) {
+      if (avoidBufferResultAllocAndCopy && isa<memref::AllocOp>(orig.getDefiningOp())) {
```
Just curious about this: does it support the case where the shape of the MemRef is unknown at compile time?

I mean for example like following:

```
// the shape of %arg0 and %arg1 may be different.
func.func @main(%arg0: memref<?x?xf32>, %arg1 : memref<?x?xf32>) -> memref<?x?xf32> {
  %0 = scf.if %pred {
    scf.yield %arg0 :  memref<?x?xf32>
  } else {
    scf.yield %arg1 :  memref<?x?xf32>
  }
  %2 = memref.alloc(…)
  concat(%0, %0, %2)
  return %2 : memref<?x?xf32>
}
```

->

```
// the shape of %2 is depends on the runtime value, how does the user to alloc the buffer in advance and pass it to the funciton?
func.func @main(%arg0: memref<?x?xf32>, %arg1 : memref<?x?xf32>, %2 : memref<?x?xf32>) {
  %0 = scf.if %pred {
    scf.yield %arg0 :  memref<?x?xf32>
  } else {
    scf.yield %arg1 :  memref<?x?xf32>
  }
  concat(%0, %0, %2)
  return
}
```

Menooker replied:
Thanks for the comments. This added option in the pass was supposed to work on statically shaped allocated memref. I will update the code to filter out dynamic-shaped cases.

matthias-springer commented on mlir/include/mlir/Dialect/Bufferization/Transforms/Passes.td:
```diff
@@ -320,6 +320,11 @@ def BufferResultsToOutParams : Pass<"buffer-results-to-out-params", "ModuleOp">
        "Add the attribute 'bufferize.result' to all output parameters.">,
+    Option<"avoidBufferResultAllocAndCopy", "avoid-buffer-result-alloc-copy",
+       "bool", /*default=*/"false",
+       "When the returned memref has static shape and is allocated by "
```
This description is rendered as a table (I think when running `mlir-opt -help`) and usually just one line.

How about renaming this option to `hoist-static-allocs`? And as a description for the option: `"Hoist static allocations to call sites."` (and moving the current description to the pass description above).

Menooker replied:
Thanks for the suggestions! I have updated the code.

ciyongch replied:
This enhancement is indeed helpful to remove the overhead of result copy.
Just curious about the new name:  as the original pass always requires the the caller to allocate the buffer for the output params(newly added function argument), it might be a bit confusing that we're running with "buffer-results-to-out-params" pass but with "hoist-static-allocs=0".

matthias-springer replied:
A new allocation is indeed always placed at the call site. But I would call this "hoisting" only if an equivalent allocation is removed in the body of the callee. (Hoisting = Moving) This happens only if `hoist-static-allocs=1`.

ciyongch replied:
I see, thanks for the clarification:)

Menooker committed:
[MLIR][Bufferization] BufferResultsToOutParams: Add an option to eliminate AllocOp and avoid Copy

Add an option hoist-static-allocs to remove the unnecessary memref.alloc and memref.copy after this pass, when the memref in ReturnOp is allocated by memref.alloc and is statically shaped. Instead, it replaces the uses of the allocated memref with the memref in the out argument.
By default, BufferResultsToOutParams will result in a memcpy operation to copy the originally returned memref to the output argument memref. This is inefficient when the source of memcpy (the returned memref in the original ReturnOp) is from a local AllocOp. The pass can use the output argument memref to replace the locally allocated memref for better performance. elim-alloc-copy avoids dynamic allocation and memory movement.
This option will be critical for performance-sensivtive applications, which require BufferResultsToOutParams pass for a caller-owned output buffer calling convension.

ZhennanQin reviewed (approved):
LGTM.
